common/v2/features/CreateWallet/Keystore/components/VerifyKeystorePanel.tsx

```diff
@@ -0,0 +1,201 @@
+
+const UploadZone = styled.label`
+  width: 100%;
+  font-size: 13px;
+  text-transform: uppercase;
+  border: solid 1px #e5ecf3;
+  background-color: rgba(247, 247, 247, 0.4);
+  word-wrap: break-word;
+  padding: 16px 4px;
+  margin-top: 8px;
+  text-align: center;
+  cursor: pointer;
+`;
+
+const HiddenUploadZone = styled.input`
+  &[type='file'] {
+    display: none;
+  }
+`;
+
+const ErrorWrapper = styled.div`
+  margin-top: 26px;
+`;
+
+interface Props extends PanelProps {
+  privateKey: string;
+  verifyKeystore(keystore: string, password: string): Promise<boolean>;
+  verifyPrivateKey(key: string, password: string): boolean;
+  addCreatedAccountAndRedirectToDashboard(): void;
+}
+
+class VerifyKeystorePanel extends Component<Props> {
+  public state = {
+    validating: false,
+    submited: false,
+    keystore: null,
+    password: '',
+    privateKey: '',
+    fileName: '',
+    passwordError: '',
+    privateKeyError: '',
+    emptyFormError: false
+  };
+
+  public validating = false;
+
+  public validate = async () => {
+    const {
+      verifyKeystore,
+      verifyPrivateKey,
+      privateKey: generatedPrivateKey,
+      addCreatedAccountAndRedirectToDashboard
+    } = this.props;
+    const { keystore, password, privateKey } = this.state;
+    this.validating = true;
+
+    this.setState({ passwordError: '', privateKeyError: '', emptyFormError: false });
+
+    if (!keystore && !privateKey) {
+      this.setState({ emptyFormError: true });
+      return;
+    }
+
+    if (keystore) {
+      const isValid = await verifyKeystore(keystore, password);
+      if (!isValid) {
+        this.setState({ passwordError: 'Wrong password' });
+        return;
+      } else {
+        addCreatedAccountAndRedirectToDashboard();
+      }
+    } else if (privateKey) {
+      const isValid = verifyPrivateKey(privateKey, password) && generatedPrivateKey === privateKey;
+      if (!isValid) {
+        this.setState({ privateKeyError: 'Invalid private key' });
+        return;
+      } else {
+        addCreatedAccountAndRedirectToDashboard();
+      }
+    }
+    this.setState({ submited: true });
+    this.validating = false;
+  };
+
+  public handleFileSelection = (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (!e.target || !e.target.files) {
+      return;
+    }
+
+    const fileReader = new FileReader();
+    fileReader.readAsText(e.target.files[0], 'UTF-8');
+    this.setState({ fileName: e.target.files[0].name });
+
+    fileReader.onload = () => {
+      this.setState({ keystore: fileReader.result });
+    };
+  };
+
+  public handlePasswordInputChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
+    this.setState({ password: e.target.value });
+  };
+
+  public handlePrivateKeyInputChanged = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
+    this.setState({ privateKey: e.target.value });
+  };
+
+  public render() {
+    const { onBack, totalSteps, currentStep } = this.props;
+
+    return (
+      <ExtendedContentPanel
```

Comment:
Thank you for the suggestion. With this approach, there is still some redundant code for each flow. I agree we should use a common pattern across the project. Maybe we can come up with something a little more reusable since there is a lot of flows going on in the application.